src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -1907,6 +1907,9 @@ void PhaseIdealLoop::do_unroll(IdealLoopTree *loop, Node_List &old_new, bool adj
         "odd trip count for maximally unroll");
     // Don't need to adjust limit for maximally unroll since trip count is even.
   } else if (loop_head->has_exact_trip_count() && init->is_Con()) {
+    // We should not be here if we have old_trip_count == max_juint
+    // it would make trip_count == 2^31 which causes overflow and the situation is overall weird
```

Comment:
Can you say something a little more specific than "weird"? As a reader, it's not immediately clear what that may imply.

Reply:
I've tried to explain the concerns better but tbh, it's also part of the weirdness: it's not clear how inconsistent the state would be and what would and should happen.